lib/launcher.ts

```diff
@@ -108,197 +107,160 @@ let initFn = function(configFile: string, additionalConfig: Config) {
-                  cleanUpAndExit(RUNNERS_FAILED_EXIT_CODE);
-                });
+  const totalTasks = scheduler.numTasksOutstanding();
+  let forkProcess = false;
+  if (totalTasks > 1) {  // Start new processes only if there are >1 tasks.
+    forkProcess = true;
+    if (config.debug) {
+      throw new ConfigError(
+          logger, 'Cannot run in debug mode with multiCapabilities, count > 1, or sharding');
+    }
+  }
+
+  const createNextTaskRunner = async () => {
+    return new Promise(async (resolve) => {
+      const task = scheduler.nextTask();
+      if (task) {
+        const taskRunner = new TaskRunner(configFile, additionalConfig, task, forkProcess);
+        try {
+          const result = await taskRunner.run();
+          if (result.exitCode && !result.failedCount) {
+            logger.error('Runner process exited unexpectedly with error code: ' + result.exitCode);
           }
-        };
-        // Start `scheduler.maxConcurrentTasks()` workers for handling tasks in
-        // the beginning. As a worker finishes a task, it will pick up the next
-        // task
-        // from the scheduler's queue until all tasks are gone.
-        for (let i = 0; i < scheduler.maxConcurrentTasks(); ++i) {
+          taskResults_.add(result);
+          task.done();
           createNextTaskRunner();
+          // If all tasks are finished
+          if (scheduler.numTasksOutstanding() === 0) {
+            resolve();
+          }
+          logger.info(scheduler.countActiveTasks() + ' instance(s) of WebDriver still running');
+        } catch (err) {
+          const errorCode = ErrorHandler.parseError(err);
+          logger.error('Error:', (err as any).stack || err.message || err);
+          await cleanUpAndExit(errorCode ? errorCode : RUNNERS_FAILED_EXIT_CODE);
         }
-        logger.info('Running ' + scheduler.countActiveTasks() + ' instances of WebDriver');
+      }
+    });
+  };
+
+  const maxConcurrentTasks = scheduler.maxConcurrentTasks();
+  for (let i = 0; i < maxConcurrentTasks; ++i) {
+    await createNextTaskRunner();
+  }
+  logger.info('Running ' + scheduler.countActiveTasks() + ' instances of WebDriver');
 
-        // By now all runners have completed.
-        deferred.promise
-            .then(function() {
-              // Save results if desired
-              if (config.resultJsonOutputFile) {
-                taskResults_.saveResults(config.resultJsonOutputFile);
-              }
+  // By now all runners have completed.
+  // Save results if desired
+  if (config.resultJsonOutputFile) {
+    taskResults_.saveResults(config.resultJsonOutputFile);
+  }
+
+  taskResults_.reportSummary();
+  let exitCode = 0;
+  if (taskResults_.totalProcessFailures() > 0) {
+    exitCode = RUNNERS_FAILED_EXIT_CODE;
+  } else if (taskResults_.totalSpecFailures() > 0) {
+    exitCode = 1;
+  }
+  await cleanUpAndExit(exitCode);
+  // Start `scheduler.maxConcurrentTasks()` workers for handling tasks in
```

Comment:
This comment should be up by `const maxConcurrentTasks`.

Reply:
Done.